src/vote_graph.rs

```diff
@@ -0,0 +1,515 @@
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with finality-afg. If not, see <http://www.gnu.org/licenses/>.
+
+//! Maintains the vote-graph of the blockchain.
+//!
+//! See docs on `VoteGraph` for more information.
+
+use std::collections::{HashMap, HashSet};
+use std::fmt::Debug;
+use std::hash::Hash;
+
+use super::{Chain, Error};
+
+#[derive(Debug)]
+struct Entry<H, V> {
+	number: usize,
+	// ancestor hashes in reverse order, e.g. ancestors[0] is the parent
+	// and the last entry is the hash of the parent vote-node.
+	ancestors: Vec<H>,
+	descendents: Vec<H>, // descendent vote-nodes
+	cumulative_vote: V,
+}
+
+impl<H: Hash + PartialEq + Clone, V> Entry<H, V> {
+	// whether the given hash, number pair is a direct ancestor of this node.
+	// `None` signifies that the graph must be traversed further back.
+	fn in_direct_ancestry(&self, hash: &H, number: usize) -> Option<bool> {
+		self.ancestor_block(number).map(|h| h == hash)
+	}
+
+	// Get ancestor block by number. Returns `None` if there is no block
+	// by that number in the direct ancestry.
+	fn ancestor_block(&self, number: usize) -> Option<&H> {
+		if number >= self.number { return None }
+		let offset = self.number - number - 1;
+
+		self.ancestors.get(offset)
+	}
+
+	// get ancestor vote-node.
+	fn ancestor_node(&self) -> Option<H> {
+		self.ancestors.last().map(|x| x.clone())
+	}
+}
+
+/// Maintains a DAG of blocks in the chain which have votes attached to them,
+/// and vote data which is accumulated along edges.
+pub struct VoteGraph<H: Hash + Eq, V> {
+	entries: HashMap<H, Entry<H, V>>,
+	heads: HashSet<H>,
+	base: H,
+}
+
+impl<H, V> VoteGraph<H, V> where
+	H: Hash + Eq + Clone + Ord,
+	V: ::std::ops::AddAssign + Default + Clone,
+{
+	fn new(base_hash: H, base_number: usize) -> Self {
+		let mut entries = HashMap::new();
+		entries.insert(base_hash.clone(), Entry {
+			number: base_number,
+			ancestors: Vec::new(),
+			descendents: Vec::new(),
+			cumulative_vote: V::default(),
+		});
+
+		let mut heads = HashSet::new();
+		heads.insert(base_hash.clone());
+
+		VoteGraph {
+			entries,
+			heads,
+			base: base_hash,
+		}
+	}
+
+	/// Insert a vote with given value into the graph at given hash and number.
+	pub fn insert<C: Chain<H>>(&mut self, hash: H, number: usize, vote: V, chain: &C) -> Result<(), Error> {
+		match self.find_containing_nodes(hash.clone(), number) {
+			Some(containing) => if containing.is_empty() {
+				self.append(hash.clone(), number, chain)?;
+			} else {
```

Comment:
Shouldn't this be indented one level more rather than be on the level as the `match`-arms?

Reply:
I don't think so but open for discussion